ch_7_Finetune/7_1_Finetune_Llama2-7B.md

```diff
@@ -0,0 +1,321 @@
+
+To verify if the LoRA weights have worked in conjunction with the pretrained model, the first layer weights (which in llama2 case are trainable queries) are extracted to highlight the difference.
+
+```python
+first_weight = base_model.model.layers[0].self_attn.q_proj.weight
+first_weight_old = first_weight.clone()
+lora_weight = lora_model.base_model.model.model.layers[0].self_attn.q_proj.weight
+assert torch.allclose(first_weight_old, first_weight)
+```
+With the new merging method `merge_and_unload`, we can easily combine the fine-tuned model with pre-trained model, and testify whether the weights have changed with the `assert` statement. 
+
+```python
+lora_model = lora_model.merge_and_unload()
+lora_model.train(False)
+assert not torch.allclose(first_weight_old, first_weight)
+```
+You may get the outputs below without error report to indicate the successful conversion.
+```
+Using pad_token, but it is not set yet.
+Using pad_token, but it is not set yet.
+```
+Finally we can save the fine-tuned model in a specified local path (in our case is `./outputs/checkpoint-200-merged`).
+```python
+output_path = ./outputs/checkpoint-200-merged
+lora_model_sd = lora_model.state_dict()
+deloreanized_sd = {
+        k.replace("base_model.model.", ""): v
+        for k, v in lora_model_sd.items()
+        if "lora" not in k
+    }
+base_model.save_pretrained(output_path, state_dict=deloreanized_sd)
+tokenizer.save_pretrained(output_path)
+
+```
+
+
+## 7.1.4 Inference with Fine-tuned model
+After merging and deploying the models, we can test the performance of the fine-tuned model. 
+The detailed instructions of running
+LLM inference with BigDL-LLM optimizations could be found in [Chapter 6](../ch_6_GPU_Acceleration/6_1_GPU_Llama2-7B.md), here we quickly go through the preparation of model inference.
+
+### 7.1.4.1 Inference with the Fine-tuned Model
+```python
+model_path = "./outputs/checkpoint-200-merged"
+model = AutoModelForCausalLM.from_pretrained(pretrained_model_name_or_path = model_path,load_in_4bit=True)
+model.config.use_cache = True
+model = model.to('xpu')
+tokenizer = LlamaTokenizer.from_pretrained(pretrained_model_name_or_path = model_path)
+```
+> **Note**
+> The `model_path` argument should be consistent with the output path of your merged model.
+>
+Then we can verify if the fine-tuned model can produce reasonable and philosophical response with the new dataset added.
+```python
+with torch.inference_mode():
+    input_ids = tokenizer.encode('The paradox of time and eternity is', 
+    return_tensors="pt").to('xpu')
+    output = model.generate(input_ids, max_new_tokens=32)
+    output_str = tokenizer.decode(output[0], skip_special_tokens=True)
+    print(output_str)
+```
+Here are the inference results: 
+```
+The paradox of time and eternity is that, on the one hand, we experience time as linear and progressive, and on the other hand, we experience time as cyclical. And the
+```
+
+### 7.1.4.2 Inference with the Pre-trained Model
```

Comment:
We could directly show the difference between the results of fine-tuned model and the pretrained model. No need to show the inference code again for the pretrained model. It may confuse the reader :)